diff --git a/rules/proto_compile.bzl b/rules/proto_compile.bzl
index c2692bf..b399e95 100644
--- a/rules/proto_compile.bzl
+++ b/rules/proto_compile.bzl
@@ -205,7 +205,18 @@ def _proto_compile_impl(ctx):
         ### Part 2.3: build --{name}_out=OPTIONS argument
 
         # mut <string>
-        out = plugin.out
+        strip_prefixes = [
+            "rt/proto/3rdparty/rajant/protos",
+            "rt/proto/libs/libbdican",
+            "rt/proto/public_api/bosdyn/protos",
+            "rt/proto/libs/libtwoloop",
+            "rt/proto/libs/libprotos",
+        ]
+        pkg = ctx.label.package
+        for s in strip_prefixes:
+            if pkg.startswith(s):
+                pkg = s
+        out = "/".join([plugin.out, pkg])
         if ctx.label.workspace_root:
             out = "/".join([out, ctx.label.workspace_root])
 
